### DISCUSSION TITLE
d.ts file generation

### DISCUSSION THREAD

**Discussion by jacoobes** [squint-cljs/cherry#112] [Ideas]
hey, just curious, does the cherry compiler have d.ts file generation? 

wondering if it would be worth implementing, or the user would be better off writing d.ts files themselves

**Comment by borkdude** [comment]
it doesn't, how would this work?

**Comment by jacoobes** [comment]
d.ts files are ways to provide intellisense for better develop experience when importing npm packages. As cherry emits ES6 javascript, typescript consumers would need d.ts files to import it into their own projects. the typescript compiler can generate d.ts files as an option. it resembles .h files in c projects in a way. Would be a nice feature, in my opinion

<https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html>, a reference

**Comment by borkdude** [comment]
clojure doesn't have a type system, I don't know what you expect cherry to do exactly. can you give some concrete examples?

Recently somebody mentioned that you can insert JSDoc comments in regular .js files and that some babel plugin can pick up on this. Would this what you mean? I'm converting this to a discussion, since it's pretty vague still what we're talking about here.

**Reply by jacoobes** [reply]
I don't have any concrete examples off the bat. The jsdoc approach seems really nice. Im not the most experienced in clojure, but i recall its macro system. I recall a project in rust ( why can i not find it anymore ) auto generated d.ts files with macros. It looked similar to this
```rs
#[define_type = "declare function foo(): string"]
fn foo() -> JsValue<String> {


}
```
 I wonder if its possible to use clojure macros to declare the shape / type of a def / defn / struct, similar to that

**Reply by martinklepsch** [reply]
One way of achieving this could be through a schema library like Malli, which then would need to support exporting of JS types. Some work has been done in this direction: https://github.com/flowyourmoney/malli-ts
